Tests/Tests/STPBankAccountTest.m

```diff
@@ -100,53 +117,82 @@ - (void)testBankAccountEquals {
-    STPBankAccount *bankAccountWithAttributes = [STPBankAccount decodedObjectFromAPIResponse:apiResponse];
-
-    XCTAssertEqualObjects([bankAccountWithAttributes bankAccountId], @"something", @"bankAccountId is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes last4], @"6789", @"last4 is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes bankName], @"STRIPE TEST BANK", @"bankName is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes country], @"US", @"country is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes fingerprint], @"something", @"fingerprint is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes currency], @"usd", @"currency is set correctly");
-    XCTAssertEqual([bankAccountWithAttributes status], STPBankAccountStatusNew);
-    XCTAssertEqualObjects([bankAccountWithAttributes accountHolderName], @"John Doe");
-    XCTAssertEqual([bankAccountWithAttributes accountHolderType], STPBankAccountHolderTypeCompany);
-
-    NSDictionary *allResponseFields = bankAccountWithAttributes.allResponseFields;
-    XCTAssertEqual(allResponseFields[@"foo"], @"bar");
-    XCTAssertEqual(allResponseFields[@"last4"], @"6789");
-    XCTAssertNil(allResponseFields[@"baz"]);
+- (void)testDecodedObjectFromAPIResponseRequiredFields {
+    NSArray<NSString *> *requiredFields = @[
+                                            @"id",
+                                            @"last4",
+                                            @"bank_name",
+                                            @"country",
+                                            @"currency",
+                                            @"status",
+                                            ];
+
+    for (NSString *field in requiredFields) {
+        NSMutableDictionary *response = [[self completeAttributeDictionary] mutableCopy];
+        [response removeObjectForKey:field];
+
+        XCTAssertNil([STPBankAccount decodedObjectFromAPIResponse:response]);
+    }
+
+    XCTAssert([STPBankAccount decodedObjectFromAPIResponse:[self completeAttributeDictionary]]);
+}
+
+- (void)testDecodedObjectFromAPIResponseMapping {
+    NSDictionary *response = [self completeAttributeDictionary];
+    STPBankAccount *bankAccount = [STPBankAccount decodedObjectFromAPIResponse:response];
+
+    XCTAssertEqualObjects(bankAccount.bankAccountId, @"ba_1AXvnKEOD54MuFwSotKc6xq0");
+    XCTAssertEqualObjects(bankAccount.accountHolderName, @"Jane Austen");
+    XCTAssertEqual(bankAccount.accountHolderType, STPBankAccountHolderTypeIndividual);
+    XCTAssertEqualObjects(bankAccount.bankName, @"STRIPE TEST BANK");
+    XCTAssertEqualObjects(bankAccount.country, @"US");
+    XCTAssertEqualObjects(bankAccount.currency, @"usd");
+    XCTAssertEqualObjects(bankAccount.fingerprint, @"C5fW7AwE3of8bHvV");
+    XCTAssertEqualObjects(bankAccount.last4, @"6789");
+    XCTAssertNil(bankAccount.routingNumber);
```

Comment:
Yeah it is super weird – I _think_ the history there is that previously `STPCard` was used as both the params object and the returned object, and we wanted to introduce `STPCardParams` with a deprecation, rather than a breaking change.

I think the best fix here would be to make `STPBankAccount` not a subclass of `STPBankAccountParams` and do the same with `CardParams` – maybe we should save that for a separate PR though.